### ISSUE TITLE
Custom webhook fails action depending on payload.

### ISSUE THREAD

**Issue by dennybaa** [StackStorm/st2#1368]
Hi there, I've faced a weird issue.

Say it we have a webhook action like this: https://gist.github.com/dennybaa/00d1936af2f441cfff15
If I issue the following commands:
1. `curl -X POST --data "{\"message\": \"6fd781b849bb\"}" http://172.17.0.8:9101/v1/webhooks/sample`
2. `curl -X POST --data "{\"hits.rate_15m\": 0.4765927732670934, \"hits.count\": 30, \"tags\": [\"metric\"], \"@timestamp\": \"2015-04-23T13:58:28.319Z\", \"hits.rate_5m\": 0.31123263029226506, \"@version\": \"1\", \"message\": \"6fd781b849bb\", \"hits.rate_1m\": 0.08765952620175455}"`

In both cases trigger is dispatched. But the action takes place only in the #1 case.

**Comment by jfryman** [comment]
I can successfully reproduce this. `0.8.3`

**Comment by Kami** [comment]
Thanks for the report.

Are we sure it's action not getting executed and not action getting executed and failing (e.g. echo failing because of unescaped special characters)?

In any case, I can dig in later.

**Comment by dennybaa** [comment]
I was thinking about it, but st2 passes json and it should take care of escaping of `{{trigger.body}}`.
As of bash it self, echo is not failing http://gyazo.com/a369b2b3ea87e9618622f802fcc978bb.

Anyway in the rulesengine log there should be an AUDIT action info about failed or executed action.

**Comment by lakshmi-kannan** [comment]
I am looking into this.

**Comment by jfryman** [comment]
https://gist.github.com/jfryman/e932eeb74f9b740dacf1

**Comment by lakshmi-kannan** [comment]
Weird finding! If there is a '.' character in the field name, the rule doesn't work. 

```
~/st2 git:master ❯❯❯ curl -X POST --data "{\"message\": \"6fd781b849bb\"}" http://localhost:9101/v1/webhooks/sample                                                                                                                                  ✭ ✖ ✱ ◼
{"message": "6fd781b849bb"}%                                                                                                                                                                                                                                 ~/st2 git:master ❯❯❯ curl -X POST --data "{\"message\": \"6fd781b849bb\", \"pi\": 3.14}" http://localhost:9101/v1/webhooks/sample                                                                                                                    ✭ ✖ ✱ ◼
{"message": "6fd781b849bb", "pi": 3.14}%                                                                                                                                                                                                                     ~/st2 git:master ❯❯❯ curl -X POST --data "{\"message\": \"6fd781b849bb\", \"pi.dot\": 3.14}" http://localhost:9101/v1/webhooks/sample                                                                                                                ✭ ✖ ✱ ◼
{"pi.dot": 3.14, "message": "6fd781b849bb"}%                                                                                                                                                                                                                 ~/st2 git:master ❯❯❯
```
```
~/st2 git:master ❯❯❯ cat /tmp/st2.webhook_sample.out                                                                                                                                                                                                 ✭ ✖ ✱ ◼
{u'message': u'6fd781b849bb'}
{u'message': u'6fd781b849bbc'}
{u'message': u'6fd781b849bbc'}
{u'message': u'6fd781b849bbc'}
{u'message': u'6fd781b849bb'}
{u'message': u'6fd781b849bb', u'pi': 3.14}
~/st2 git:master ❯❯❯
```

**Comment by lakshmi-kannan** [comment]
Yeah, confirmed. Changing `.` to `_` makes it work.

```
~/st2 git:master ❯❯❯ cat /tmp/st2.webhook_sample.out                                                                                                                                                                                               ⏎ ✭ ✖ ✱ ◼
{u'message': u'6fd781b849bb'}
{u'message': u'6fd781b849bbc'}
{u'message': u'6fd781b849bbc'}
{u'message': u'6fd781b849bbc'}
{u'message': u'6fd781b849bb'}
{u'message': u'6fd781b849bb', u'pi': 3.14}
{u'hits_rate_15m': 0.4765927732670934, u'tags': [u'metric'], u'hits_rate_5m': 0.31123263029226506, u'@timestamp': u'2015-04-23T13:58:28.319Z', u'hits_count': 30, u'hits_rate_1m': 0.08765952620175455, u'message': u'6fd781b849bb', u'@version': u'1'}
~/st2 git:master ❯❯❯
````

**Comment by lakshmi-kannan** [comment]
We have been `pickled`.

**Comment by lakshmi-kannan** [comment]
Long story short, we use pickle for serializing objects for rabbitmq transport. We always knew JSON was the route to be taken but used pickle as a shortcut. In this particular case, the shortcut turned out to be disastrous. We need to switch to JSON serialization and it's an involved/major change. It has ability to break functionality. We should target this for 0.10.0 release. For now, the work around is not use `_` instead of `.` in payload fields.

**Comment by Kami** [comment]
I've encountered a similar issue today when I was working on something today.

The issue is related to the payload containing dots and has nothing to do with pickle. If payload contains a dot, RulesEngine will blow up since TriggerInstance payload value can't contain dots because we don't use a special escaped dict field.

I will open a PR with fix for this shortly.

**Comment by Kami** [comment]
I've merged PR with a fix into master.

@dennybaa You can test the fix by downloading a new 0.10dev release from https://downloads.stackstorm.net/releases/st2/0.10dev/ when it's built (I've also already tested it myself and added a test case for it).
